File: go/collatz.go
//Trevor Mines
//Collatz iterative
//CSC 330
//Fall 2020

package main

import "fmt"

func main(){
	var ic int64
	var i, j int64
	i = 0
	j = 0
	largestLens := [] int64{0,0,0,0,0,0,0,0,0,0}
	topVals := [] int64{0,0,0,0,0,0,0,0,0,0}	

	//Loops through each value calculating the collatz length
	for j = 2; j < 5000000000; j++{
		i = j 
		ic = 0
		for i > 1{
			if i%2 == 0{
				i /= 2
			}else{
				i = 3 * i + 1
			}
			ic++
		}

	//Puts the lengths into largestLens if it is larger than the smallest length in largestLens
	//Puts the value into topVals and sorts
	if ic > largestLens[0]{
		if !contains(largestLens, ic){	
			largestLens[0] = ic
			topVals[0] = j
			bubblesort(largestLens, topVals)	
		}
	}


	

	} 


	//Prints the output in the specified format, sorted by length then value resorting in between 
	fmt.Println("Sorted based on ascending length:")
	for l := 0; l < 10; l++{
		fmt.Println("A starting value of ", topVals[l], " has ", largestLens[l], " steps.")
	}

	bubblesort(topVals, largestLens)
	fmt.Println()

	fmt.Println("Sorted based on ascending value:")
	for l := 0; l < 10; l++{
                fmt.Println("A starting value of ", topVals[l], " has ", largestLens[l], " steps.")
        }


}

//Sorts array a and makes the same swaps in array b to keep the numbers associated with eachother
func bubblesort(a []int64, b []int64) {
    for itemCount := len(a) - 1; ; itemCount-- {
        hasChanged := false
        for index := 0; index < itemCount; index++ {
            if a[index] > a[index+1] {
                a[index], a[index+1] = a[index+1], a[index]
		b[index], b[index+1] = b[index+1], b[index]
                hasChanged = true
            }
        }
        if hasChanged == false {
            break
        }
    }
}

//Searches the passed in array for an item and returns a bool based on if it is found
func contains(array []int64, item int64) bool {
	for k := 0; k < 10; k++ {
		if array[k] == item{
			return true
		}
	}
	return false
}

File: julia/collatz.jl
#!/usr/bin/julia

#Trevor Mines
#Collatz iterative
#CSC 330
#Fall 2020

#Function that sorts the first array and makes the same swaps in the second array
function bubblesort(largestLens::AbstractVector, topVals::AbstractVector)
        for _ in 2:length(largestLens), j in 1:length(largestLens)-1
                if largestLens[j] > largestLens[j+1]
                        largestLens[j], largestLens[j+1] = largestLens[j+1], largestLens[j]
                        topVals[j], topVals[j+1] = topVals[j+1], topVals[j]
                end
        end
end

function main()

	largestLens :: Array{Int64,1} = [0,0,0,0,0,0,0,0,0,0]
	topVals :: Array{Int64,1} = [0,0,0,0,0,0,0,0,0,0]

	#Loops through each value and gets the ic	
	j :: Int64 = 2
	while j < 5000000000
		ic = 0
		i :: Int64 = j
		while i != 1
			if i % 2 == 0
				i = i ÷ 2
			else 
				i = i * 3 + 1
			end
			ic = ic + 1
		end
	
		#Puts the length in largestLens if it is larger than the smallest value in largestLens
		#Mirrors changes in topVals and sorts
		if ic > largestLens[1]
			if (indexin(ic, largestLens)) .== nothing
				largestLens[1] = ic
				topVals[1] = j
				bubblesort(largestLens,topVals)
			end
		end

		j = j + 1

	end

	#Prints in the specified format, sorted first by lengths then by values
	println("Sorted based on ascending length:")
        ic = 1
        while ic <= 10
                println("A starting value of ", topVals[ic], " has ", largestLens[ic], " steps.")
                ic = ic + 1
        end

        bubblesort(topVals, largestLens)

        println()
        println("Sorted based on ascending value:")

        ic = 1
        while ic <= 10
                println("A starting value of ", topVals[ic], " has ", largestLens[ic], " steps.")
                ic = ic + 1
        end

end


main()

exit(0)
